src/test/scala/io/citrine/lolo/DataGenerator.scala

```diff
@@ -8,33 +8,32 @@ import io.citrine.random.Random
 object DataGenerator {
 
   /** Sequence of inputs and labels generated for training a learner. */
-  case class TrainingData[I, L](data: Seq[(Vector[I], L)])
-
-  // An implicit class is used to restrict binning to when the inputs are of type Double
-  implicit class TrainingInputBinner[L](trainingData: TrainingData[Double, L]) {
-    def withBinnedInputs(bins: Seq[(Int, Int)]): TrainingData[Any, L] = {
-      val (baseInputs, baseLabels) = trainingData.data.unzip
+  case class TrainingData[T](data: Seq[TrainingRow[T]]) {
+    def withBinnedInputs(bins: Seq[(Int, Int)]): TrainingData[T] = {
+      val baseInputs = data.map(_.inputs)
       val indexToBins = bins.toMap
       val transposedInputs = baseInputs.transpose
       val binnedInputs = Vector
         .tabulate(transposedInputs.length) { index =>
-          val indexData = transposedInputs(index)
+          val indexData = transposedInputs(index).asInstanceOf[Seq[Double]]
```

Comment:
We have to cast here now when binning data for various unit tests, but its exclusively used within the test code so :shrug:

Reply:
You could also create a `RealValuedTrainingData` class that enforced the inputs as `Double`. But then it would have to have some function to cast to the `TrainingRow` type and you'd be calling that function dozens of times across many tests.
I think that this single `asInstanceOf` is a more reasonable price to pay.